### PR TITLE
fixed weights margin example added

### PR DESCRIPTION
**Pull request by andrewcz**
create a pull request with my attempt.

### CODE REVIEW

File: examples/margin_loan_jpy_example.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import time
+import matplotlib.pyplot as plt
+import pandas as pd
+import cvxportfolio as cvx
+
+# Define a subclass of DownloadedMarketData to rebase asset returns in JPY
+class JPYRebasedMarketData(cvx.DownloadedMarketData):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.exchange_rate_to_jpy = self.get_exchange_rate_to_jpy()
```
Comment (enzbus) [inline examples/margin_loan_jpy_example.py:10]: Here we should add 
```
self.returns = cvxportfolio.utils.set_pd_read_only((1 + self.returns).multiply((1 + self.exchange_rate_to_jpy.pct_change()), axis=0) - 1)
self.prices = ...
self.volumes = ... # original is in USD, transform to JPY
```

**Comment by enzbus** [comment]
I tried to do correct time-alignment of FX conversion and that's what I have (might still be off, but I don't think by much):

```
import cvxportfolio as cvx
import matplotlib.pyplot as plt
import numpy as np

# AAPL is timestamped 9:30 NY time
aapl = cvx.YahooFinance('AAPL').data

# rate is timestamped 0:00 UTC (~midnight London)
jpy_usd = cvx.YahooFinance('JPY=X').data

# take close from day before rather than open, seems cleaner
jpy_usd_rate = jpy_usd.close.shift(1)

# reindex, taking last available one
jpy_usd_rate = jpy_usd_rate.reindex(aapl.index, method='ffill')

# ccy returns, open-to-open NY time
jpy_usd_return = jpy_usd_rate.pct_change().shift(-1)

# aapl open-to-open total returns with JPY as base ccy
aapl_returns_in_jpy = (1 + aapl['return']) * (1 + jpy_usd_return) - 1


np.log(1 + aapl['return']).cumsum().plot(label='AAPL cum logret, USD')
np.log(1 + aapl_returns_in_jpy).cumsum().plot(label='AAPL cum logret, JPY')
plt.legend()
plt.show()
```
Screenshot from 2024-03-03 14-54-51

**Comment by andrewcz** [comment]
amazing!!! thank you looks pretty spot on.
I'm not sure this is relevant but i was thinking of adding a margin call line (estimating maintenance margin).
the portfolio of adding other assets with fixed weights should be fairly trivial.

**Comment by enzbus** [comment]
Re: margin call. For that we do need to subclass `cvx.FixedWeights`. The default does not trade on days for which there are no trades provided, instead it should check the current margin and do something if it's too large. If you do provide target weights for a given day then there's no issue with margin, it's given by your target. For example if your target weights are `pd.Series({'AAPL':1, 'GOOG:1, 'JPYEN':-1})` has leverage 2 (and you tell me how you want to define margin).

Re: measuring in JPY and USD. For that also `BacktestResult` needs subclassing, and I'm afraid it's not going to be as clean as the rest (I haven't finished de-coupling `MarketSimulator` and `BacktestResult`), but we can do it.

**Comment by enzbus** [comment]
Also, can you please write a few lines in the docstring of the example to explain what it does? Would also help me to help you.

**Comment by andrewcz** [comment]
Thank you @enzbus !! I will keep working on it and adding to the example.
I should be able to do it within the week :).
Best and take care.
Andrew

**Comment by enzbus** [comment]
Ok @andrewcz , let me have a look, later tonight or tomorrow.

**Comment by enzbus** [comment]
Interesting @andrewcz , I had a quick look and will make a pass. Sounds like that custom cost object can be neatly represented by a `TransactionCost` subclass. Regarding interest rate cost, right now it's applying the default central bank rate, which is this series https://fred.stlouisfed.org/series/INTDSRJPM193N . Currently it's 0.3% annual. I saw somewhere you were hard-coding 1% . We should also add a custom holding cost that applies the spread (on negative cash only). Then I think it's done, a full example of many custom terms with relatively little code :)

**Comment by enzbus** [comment]
I meant this https://fred.stlouisfed.org/series/IRSTCB01JPM156N , it's defined in the RATES constant in market_data.py

**Comment by enzbus** [comment]
Another thing, I took out the margin call clause because it was already triggered by the max_leverage clause. Makes sense? Not sure if needed to have two different ones...

**Comment by andrewcz** [comment]
Legend! Amazing @enzbus !! Thank you so much for your help! It’s Currently 220am in Sydney. Will work on it as soon as I am awake :)!
Best,
Andrew

**Comment by andrewcz** [comment]
Hi @enzbus i have added some sublcasses under costs however im getting an error now running the example. are you able to assist. 
Also is it easy to plot the portfolio value in usd (as well as jpy?)

**Comment by enzbus** [comment]
Ok @andrewcz I've seen your additions, let me clean them, they should be much simpler...

**Comment by andrewcz** [comment]
> Ok @andrewcz I've seen your additions, let me clean them, they should be much simpler...

Hi @enzbus ! I hope you’re well. Apologies if I’ve missed this, have these changes been added to the project. I was going to add to the project over the weekend.
Best,
Andrew

**Comment by enzbus** [comment]
Thanks @andrewcz , I'm reviewing the code. Two points:
- Can you please merge the 3 example scripts into one? I believe only one is the final and the other two are drafts, please remove them.
- The two custom costs are OK but I can not put them into the base library (yet). The holding cost as it is defined only works for long-only portfolios (with short cash), the transaction cost is quite exotic and is better shown as an example of a custom extension.

Once that is done, only **one script in the examples**, I'll make a final pass and if you're OK with it we merge. Thanks!

**Comment by andrewcz** [comment]
Hi @enzbus ! thank you!
Im just not sure what you mean by the below 
- he two custom costs are OK (but need fixing) and I can not put them into the base library (yet). The holding cost as it is defined only works for long-only portfolios (with short cash), the transaction cost is quite exotic and is better shown as an example of a custom extension.

**Comment by enzbus** [comment]
Yes, please. And I'll fix them :)

**Comment by andrewcz** [comment]
hi @enzbus  so i have just updated the project - deleted the files and made adjustments to the example

**Comment by andrewcz** [comment]
@enzbus I just have some further background questions - 
Is it straightforward to view the specific trades executed during a portfolio rebalance, such as "bought 25 shares of Google at X price"?
How does the system handle situations where it cannot secure the algorithm's calculated price?
If I wish to incorporate additional risk metrics, where should I integrate them into the project?
Suppose I want to introduce a new indicator, such as "risk on/risk off," to adjust leverage accordingly. How feasible is this, and where would I implement it in the code?

**Comment by enzbus** [comment]
Ok @andrewcz , please make sure to not edit other files in the base library, or create other spurious files, otherwise we need to remove this PR and create a new one. Please do these things:

- Add again the docstring at the start of the script to explain what the example does;
- Write docstrings for the custom costs to explain what they do, I'll fix their code later to make sure they use the right interfaces.

For your last questions:


> @enzbus I just have some further background questions -
> 
>     1. Is it straightforward to view the specific trades executed during a portfolio rebalance, such as "bought 25 shares of Google at X price"?
 `BackTestResult` records the trade vectors `u`, in units of cash (in this case `JPYEN`) and not the number of shares traded. The trade vectors are rounded to integer number of shares (with the `round_trades` option in `MarketSimulator`). It is easy to recover the shares traded, by `result.u.iloc[:,:-1] / prices`, if you have a dataframe with the prices.

>     2. How does the system handle situations where it cannot secure the algorithm's calculated price?
Slippage is accounted for by `TransactionCost`. 

>     3. If I wish to incorporate additional risk metrics, where should I integrate them into the project?

Custom subclass of `BackTestResult`.

>     4. Suppose I want to introduce a new indicator, such as "risk on/risk off," to adjust leverage accordingly. How feasible is this, and where would I implement it in the code?

In the policy object, just like it is done now. The `values_in_time` method receives a full view of the past market data.

Please do the above, I'd like to merge this.

**Comment by enzbus** [comment]
PS Apologies for all the back-and-forths, I'm using this to improve entry points for custom subclassing (I just made the interface between simulator and result public this morning, allowing for custom result classes).

**Comment by andrewcz** [comment]
Hi @enzbus apologies for the lateness I’ve had a few work and university commitments. Will submit the changes in the next 24 hours.
Best,
Andrew

**Comment by andrewcz** [comment]
Hi @enzbus so i have added a docstring. I think the repo is pretty clean now (i think). i double checked the files before the push. 
it is now ready for your review.

**Comment by andrewcz** [comment]
@enzbus go for it! do your thing! any changes are more than welcome! have a great day.
Do you have suggestion on what to work on next :).
Best,
Andrew

**Pull request merged**